syntax = "proto3";

package mom;

// Service for Topics
service TopicService {
    rpc CreateTopic (CreateTopicRequest) returns (Response);
    rpc PublishMessage (PublishMessageRequest) returns (Response);
    rpc Subscribe (SubscribeRequest) returns (Response);
    rpc Unsubscribe (UnsubscribeRequest) returns (Response);
    rpc ListTopics (Empty) returns (ListTopicsResponse);
    rpc PullMessages (PullMessagesRequest) returns (MessagesResponse);
    rpc DeleteTopic (DeleteTopicRequest) returns (Response);
}

// Service for Queues
service QueueService {
    rpc CreateQueue (CreateQueueRequest) returns (Response);
    rpc ListQueues (Empty) returns (ListQueuesResponse);
    rpc PushMessage (PushMessageRequest) returns (Response);
    rpc PullMessage (PullMessageRequest) returns (MessageResponse);
    rpc DeleteQueue (DeleteQueueRequest) returns (Response);
}

service ReplicationService{
    rpc GetTopic (ReplicationRequest) returns (GetTopicResponse);
    rpc GetQueue (ReplicationRequest) returns (GetQueueResponse);
    rpc ImportQueue (ImportQueueRequest) returns (ImportReplicationResponse);
    rpc ImportTopic (ImportTopicRequest) returns (ImportReplicationResponse);

}

// Common messages
message Response {
    bool success = 1;
    string message = 2;
}

message Empty {}

message CreateTopicRequest {
    string topic_id = 1;
    string user = 2;
}

message PublishMessageRequest {
    string topic_id = 1;
    string content = 2;
    string sender = 3;
}

message SubscribeRequest {
    string topic_id = 1;
    string subscriber_id = 2;
}

message UnsubscribeRequest {
    string topic_id = 1;
    string subscriber_id = 2;
}

message Topic {
    string topic_id = 1;
    string autor = 2;
}

message Queue {
    string topic_id = 1;
    string autor = 2;
}

message ListTopicsResponse {
    repeated Topic topics = 1;
}

message PullMessagesRequest {
    string topic_id = 1;
    string subscriber_id = 2;
}

message MessagesResponse {
    bool success = 1;
    string message = 2;
    repeated Message data = 3;
}

message CreateQueueRequest {
    string queue_id = 1;
    string user = 2;
}

message ListQueuesResponse {
    repeated Queue queues = 1;
}

message Group {
    repeated string queues = 1;
}

message PushMessageRequest {
    string queue_id = 1;
    string content = 2;
    string sender = 3;
}

message PullMessageRequest {
    string queue_id = 1;
}

message MessageResponse {
    bool success = 1;
    string message = 2;
    Message data = 3;
}

message Message {
    string message_id = 1;
    string parent = 2;
    string content = 3;
    string sender = 4;
    string timestamp = 5;
}

message DeleteTopicRequest {
    string topic_id = 1;
    string user = 2;
}

message DeleteQueueRequest {
    string queue_id = 1;
    string user = 2;
}


message GetTopicResponse{
    bool success = 1;
    string message = 2;
    TopicReplication data = 3;
}

message GetQueueResponse{
    bool success = 1;
    string message = 2;
    QueueReplication data = 3;
}

message ReplicationRequest{
    string target = 1;
}

message TopicReplication {
    string topic_id = 1;
    string autor = 2;
    repeated MessageReplication messages = 3;
}

message QueueReplication {
    string queue_id = 1;
    string autor = 2;
    repeated MessageReplication messages = 3;
}

message MessageReplication {
    string message_id = 1;
    string parent = 2;
    string content = 3;
    string sender = 4;
    string timestamp = 5;
    repeated string sent = 6;
}

message ImportQueueRequest {
    QueueReplication queue = 1;
}

message ImportTopicRequest {
    TopicReplication topic = 1;
}

message ImportReplicationResponse {
    bool success = 1;
    string message = 2;
}